자동차 경주 2단계 - 문자열 계산기

bsgreentea opened the pull request:
## 구현 사항

Calculator
> 수식 받기 및 검증, 식 계산, 사칙연산을 수행할 수 있도록 했습니다.
> 수식 검증은 우선 공백인지만 확인하도록 했습니다.

OperatorType
> 사칙연산 연산자용으로 enum class를 만들었습니다.
> OperatorType.from() 으로 타입 확인 시 부적절한 값이 들어오면 익셉션 던지도록 했습니다.

그리고 각각에 대한 테스트를 위해 CalculatorTest와 OperatorTypeTest를 만들었습니다.

Commit:
feat: Calculator 클래스 선언 및 사칙연산을 위한 메서드 구성

Commit:
test: Calculator 클래스 사칙연산 메서드에 대한 테스트코드 작성

bsgreentea committed:
feat: 입력값과 연산자 타입 확인할 수 있는 검증 코드 추가

bsgreentea committed:
test: 입력 내용이 빈칸인지와 연산자가 올바른지에 대한 테스트 코드 추가

bsgreentea committed:
feat: 입력 식 계산 메서드 구현

bsgreentea committed:
test: 식 계산에 대한 테스트 코드 추가

bsgreentea committed:
refactor : 연산자 체크 로직과 테스트 코드 수정해라

- 연산자인지 확인하는 방법을 '한 글자이고 숫자일 때' 하는 것으로 변경함.
- OperationType > from으로 타입을 가져올 때, 정의되지 않은 값이라면 IllegalArgumentException 던지도록 수정
- 연산자 검증하는 테스트를 위해 OperatorTypeTest 생성
- 기존의 calculate 테스트는 `식 계산` 메서드가 역할을 하고 있기 때문에 제거함

Rok93 commented on src/main/kotlin/week1/calculator/Calculator.kt:
```diff
@@ -0,0 +1,64 @@
+
+class Calculator {
+
+    // 계산 시 연산자의 우선순위를 고려하지 않는다.
```
불필요한 주석이라면 제거하는게 좋을 것 같아요. 🙏

Rok93 commented on src/main/kotlin/week1/calculator/Calculator.kt:
```diff
@@ -0,0 +1,64 @@
+package week1.calculator
+
+class Calculator {
```
자동차 미션 때에는 README.md에 기능구현 목록표를 작성해보시고 기능구현목록표에 적어둔 기능단위로 커밋하며 체크하면서 미션을 진행해보시면 어떨까요? 🤗

Rok93 replied:
Calculator는 프로퍼티를 가지고있지 않네요?
그렇다면 꼭 객체로 만들어서 사용할 필요가 없는 것 같아요. 🤔

Calculator 클래스를 [object](https://kotlinlang.org/docs/object-declarations.html#object-declarations-overview)로 변경해보면 어떨까요? 😃

Rok93 commented on src/main/kotlin/week1/calculator/Calculator.kt:
```diff
@@ -0,0 +1,64 @@
+        return result
+    }
+
+    fun add(a: Int, b: Int): Int {
+        return a + b
+    }
+
+    fun subtract(a: Int, b: Int): Int {
+        return a - b
+    }
+
+    fun multiply(a: Int, b: Int): Int {
+        return a * b
+    }
+
+    fun divide(a: Int, b: Int): Int {
+        return a / b
+    }
+
+    fun checkInput(input: String?) {
+        if (input.isNullOrBlank()) throw IllegalArgumentException()
+    }
```
모두 Calculator 객체 내부에서만 사용되는 메서드로 보이는데,  isDigit 메서드처럼 접근제한자를 `private`으로 두시는게 어떨까요?

bsgreentea replied:
@Rok93 테스트코드 작성을 위해 private으로 설정하지 않았는데, 혹시 private이어도 테스트할 수 있는 방법이 있을지 궁금합니다!

Rok93 commented on src/main/kotlin/week1/calculator/Calculator.kt:
```diff
@@ -0,0 +1,64 @@
+
+    fun divide(a: Int, b: Int): Int {
+        return a / b
+    }
```
`덧셈`, `뺄셈`, `곱셈`, `나눗셈`은 모두 연산자의 행위로 볼 수 있지 않을까요? 🤔
OperatorType에게 연산을 수행하는 책임을 위임해보면 어떨까요? 🤗

Rok93 commented on src/main/kotlin/week1/calculator/Calculator.kt:
```diff
@@ -0,0 +1,64 @@
+            }
+        }
+
+        println("식 계산 결과 = [$result]")
```
불필요한 (콘솔) 출력문도 제거해주세요! 🙏

Rok93 commented on src/main/kotlin/week1/calculator/Calculator.kt:
```diff
@@ -0,0 +1,64 @@
+    // 계산 시 연산자의 우선순위를 고려하지 않는다.
+    fun calculate(input: String?): Int {
+        checkInput(input)
+        checkNotNull(input)
```
이미 위의 `checkInput` 메서드에서 null 체크를 하는데, `checkNotNull` 함수를 사용하신 이유가 있을까요? 🤔

bsgreentea replied:
널체크를 위해 checkInput 메서드를 호출했으나 이후 코드에서 추가적인 널 체크가 필요했고, 이게 싫어서 checkNotNull을 추가했습니다.
간단한 식이라면 메서드로 빼지 않고 맨 위에 추가하는 것도 괜찮을 것 같네요!

Rok93 commented on src/main/kotlin/week1/calculator/Calculator.kt:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    companion object {
+        private val digits = listOf("0", "1", "2", "3", "4", "5", "6", "7", "8", "9")
```
0 ~ 9 까지의 숫자들을 상수로 정의하시는 것을 의도하신게 아닌가 싶어요. 😃

[Kotlin Code Convention](https://kotlinlang.org/docs/coding-conventions.html#property-names)에 따르면 상수의 명명 규칙은 UPPER_CAMEL_CASE 로 명명하라고 가이드하고있어요. 😃

Rok93 commented on src/main/kotlin/week1/calculator/Calculator.kt:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    private fun isDigit(candidate: String): Boolean {
+        return digits.contains(candidate)
```
미리 숫자 상수들을 두고 숫자에 포함되는지를 확인하는 방법도 있지만, `정규식`을 이용한다거나 (한 자리의 숫자만 들어온다고 가정했을 때는)`Character.isDigit(candidate[0])` 같은 방식으로도 구현해볼 수 있겠네요. 😃

Rok93 commented on src/main/kotlin/week1/calculator/Calculator.kt:
```diff
@@ -0,0 +1,64 @@
+        val numbers = mutableListOf<Int>()
+        val operators = mutableListOf<OperatorType>()
+        input.split(' ').forEach {
+            if (it.length == 1 && !isDigit(it)) {
```
한 자리의 숫자만 연산한다는 조건이 있었나요? 🤔

Rok93 replied:
추후에 연산자가 한자리 문자열만 들어온다는 보장도 없으니 
`it.length == 1 && !isDigit(it)` 조건식을 `OperatorType.isOperatorSymbol(it)` 같은 방식으로 변경해볼 수도 있을 것 같아요. 

isOperatorSymbol 메서드 내에서는 **메서드 파라미터로 받은 값**이  OperatorType의 `value` 들을 모두 추출한 값에 포함되는지 여부를 파악하면 될 것 같아요. 😃  이러면 추후 OperatorType가 추가되더라도 코드 수정을 할 필요가 없을 것 같아요. 😆

Rok93 commented on src/main/kotlin/week1/calculator/Calculator.kt:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    fun checkInput(input: String?) {
+        if (input.isNullOrBlank()) throw IllegalArgumentException()
```
Kotlin에서 제공하는 [require 함수](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin/require.html)를 활용해볼 수도 있을 것 같아요. 😃


또한 왜 예외가 발생한건지 적절한 예외 메시지를 남겨보면 어떨까요? 👀 
예외 메시지에 어떤 인자가 들어왔는지 같은 정보도 포함되어있으면 예외 발생시 원인을 파악하는데 도움이될 것 같아요! 😉

Rok93 commented on src/test/kotlin/week1/calculator/CalculatorTest.kt:
```diff
@@ -0,0 +1,53 @@
+    fun `식 계산`() {
+        assertThat(calculator.calculate("2 + 3 * 4 / 2")).isEqualTo(10)
+        assertThat(calculator.calculate("1 + 2 + 3 + 4 + 5")).isEqualTo(15)
+    }
```
같은 테스트지만 다양한 파라미터에 대한 테스트를 진행하고 싶다면 [Junit의 Parameterized Test](https://www.baeldung.com/parameterized-tests-junit-5)를 활용해볼 수 있어요. 😃

Paramerized Test를 활용해서 테스트를 조금 더 깔끔하게 만들어보면 어떨까요? 💪

Rok93 commented on src/test/kotlin/week1/calculator/OperatorTypeTest.kt:
```diff
@@ -0,0 +1,20 @@
+
+class OperatorTypeTest {
+    @Test
+    fun `연산자 검증`() {
```
하나의 테스트 케이스에는 특수한 케이스가 아니라면 가급적 하나의 행위에 대한 검증만 수행해주세요!  🙏

두 가지 테스트 케이스(assertThatExceptionOfType 검증단, assertThatCode 검증단)로 분리하는 것이 좋을 것 같아요. 🤔

assertThatCode는 제가 잘 모르는 검증 메서드라 찾아보니 
`assertThatCode`는 **assertj에서의 정상동작 검증**
`assertDoesNotThrow`는 **Junit5의 정상동작 검증**
인 걸로 파악되네요. 😃

Rok93 commented on src/test/kotlin/week1/calculator/OperatorTypeTest.kt:
```diff
@@ -0,0 +1,20 @@
+            OperatorType.values().forEach {
+                OperatorType.from(it.value)
+            }
+        }
```
이 부분도 Paramaerized Test 에서  Enum class에 대한 테스트 부분을 참고해보시면 더 깔끔하게 테스트 코드를 변경해볼 수 있을 것 같아요. 🤔

Rok93 commented on src/main/kotlin/week1/calculator/Calculator.kt:
```diff
@@ -0,0 +1,64 @@
+    }
+
+    fun divide(a: Int, b: Int): Int {
+        return a / b
```
나눗셈의 경우에는 b가 0이면 어떻게 될까요? 🤔
이 케이스에 대한 예외처리가 필요하지 않을까요? 😉

Rok93 commented on src/main/kotlin/week1/calculator/Calculator.kt:
```diff
@@ -0,0 +1,64 @@
+
+        val numbers = mutableListOf<Int>()
+        val operators = mutableListOf<OperatorType>()
+        input.split(' ').forEach {
```
코드를 처음보는 사람이라면 `' '` 같은 값이 어떤 의미를 가지는지 한눈에 파악하기 힘들 것 같아요. 👀
상수로 추출하고 적절한 네이밍을 부여한다면 다른 동료들이 코드를 이해하는데 더 도움을 받을 수 있을 것 같아요. 😉

Rok93 reviewed (changes requested):
보성님 2단계 미션 빠르게 잘 구현 잘해주셨네요! 👍
조금 더 고민해보면 좋을 것 같은 부분이 있어서 몇몇 코멘트 남겨두었습니다. 😃
확인해서 반영해주세요. 🙏

코멘트 중에서 잘 이해가 안되거나 미션 진행하면서 어려운 부분이 있으면 언제든지 DM 주세요! 🙏

bsgreentea committed:
refactor : 리뷰 반영해라

- 연산 수행을 OperatorType에서 하도록 수정
- 각 OperatorType은 OperatorExecutor의 execute를 구현하도록 함
- Calculator를 object로 변경
- Calculator의 calculator 메서드를 process로 리네이밍함. 클래스와 중복되는 이름같아 불필요하다고 생각했음.
- 테스트 코드에 @ParameterizedTest 적용

Rok93 commented on src/test/kotlin/week1/calculator/OperatorTypeTest.kt:
```diff
@@ -0,0 +1,49 @@
+
+class OperatorTypeTest {
+
+    @Test
+    fun `더하기`() {
+        Assertions.assertThat(OperatorType.ADD.execute(1, 2)).isEqualTo(3)
+        Assertions.assertThat(OperatorType.ADD.execute(123000, 456)).isEqualTo(123456)
+    }
+
+    @Test
+    fun `빼기`() {
+        Assertions.assertThat(OperatorType.SUBTRACT.execute(1, 2)).isEqualTo(-1)
+        Assertions.assertThat(OperatorType.SUBTRACT.execute(10, 2)).isEqualTo(8)
+    }
+
+    @Test
+    fun `곱하기`() {
+        Assertions.assertThat(OperatorType.MULTIPLY.execute(1, 2)).isEqualTo(2)
+        Assertions.assertThat(OperatorType.MULTIPLY.execute(123, 456)).isEqualTo(56088)
+    }
+
+    @Test
+    fun `나누기`() {
+        Assertions.assertThat(OperatorType.DIVIDE.execute(1, 2)).isEqualTo(0)
+        Assertions.assertThat(OperatorType.DIVIDE.execute(100, 9)).isEqualTo(11)
+        Assertions.assertThat(OperatorType.DIVIDE.execute(100, 0)).isEqualTo(100)
+    }
```
`@MethodSource`를 활용하면 하나의 테스트 케이스로 변경해볼 수 있을 것 같아요. 🤗
```suggestion
    @ParameterizedTest
    @MethodSource
    fun `연산을 한다`(sut: OperatorType, a: Int, b: Int, expected: Int) {
        // when
        val actual = sut.execute(a, b)

        // then
        assertThat(actual).isEqualTo(actual)
    }

    companion object {
        @JvmStatic
        fun `연산을 한다`(): List<Arguments> {
            return listOf(
                Arguments.of(OperatorType.ADD, 1, 2, 3),
                Arguments.of(OperatorType.ADD, 123000, 456, 123456),
                Arguments.of(OperatorType.SUBTRACT, 1, 2, -1),
                Arguments.of(OperatorType.MULTIPLY, 1, 2, 2),
                Arguments.of(OperatorType.DIVIDE, 1, 2, 0),
            )
        }
    }
```

bsgreentea replied:
오.. 훨씬 깔끔하네요!

Rok93 commented on src/test/kotlin/week1/calculator/OperatorTypeTest.kt:
```diff
@@ -0,0 +1,49 @@
+
+    @ParameterizedTest
+    @ValueSource(strings = ["_", "@", "#", "+-", "++"])
+    fun `유효하지 않은 연산자`(invalidOperator: String) {
```
`연산자가 유효한 케이스`에 대한 검증도 진행해주시면 좋을 것 같아요. 😉

Rok93 commented on src/main/kotlin/week1/calculator/OperatorType.kt:
```diff
@@ -0,0 +1,36 @@
+    val value: String,
+) : OperatorExecutor {
+    ADD("+") {
+        override fun execute(a: Int, b: Int) = a + b
```
[질문](https://github.com/next-step/kotlin-racingcar/pull/1332#discussion_r1375149266)에 대한 해답을 찾으신 듯 하네요. 😃

[테스트 상태인 Private 메소드를 Public메소드로 변환시 Unit Testing은 어떻게 해야하나?](https://www.slipp.net/questions/253#answer-983) 글을 참고해보시면 도움이될 것 같습니다. 🤗

Rok93 commented on src/main/kotlin/week1/calculator/OperatorType.kt:
```diff
@@ -0,0 +1,36 @@
+    },
+    DIVIDE("/") {
+        override fun execute(a: Int, b: Int): Int {
+            return if (b == 0) {
```
b가 0인 경우에는 예외를 던지는게 조금 더 자연스럽지 않을까요? 🤔

사칙연산은 세계에 통용되는 개념이다보니 임의로 변경하는 것이 오히려 혼란을 야기할 수도 있을 것 같아요. 👀

bsgreentea replied:
0으로 나누면 지금도 ArithmeticException이 발생하고 메시지로 by zero를 보여주고 있어 그냥 a / b 만 반환해도 되겠군요

Rok93 commented on src/main/kotlin/week1/calculator/OperatorType.kt:
```diff
@@ -0,0 +1,36 @@
+        }
+
+        fun isOperatorSymbol(symbol: String): Boolean {
+            val type = OperatorType.from(symbol) ?: return false
+            return OperatorType.values().contains(type)
```
아래와 같이 변경해볼 수도 있을 것 같아요. 🤔
```suggestion
            return values().any { it.value == symbol }
```

Rok93 commented on src/main/kotlin/week1/calculator/OperatorType.kt:
```diff
@@ -0,0 +1,36 @@
+    ;
+
+    companion object {
+        fun from(value: String): OperatorType? {
```
이 또한 `value` 보다는 아래의 메서드처럼 메서드 파라미터 네이밍을 `symbol`로 변경하는 것이 더 적절해보이네요. 😃

Rok93 commented on src/main/kotlin/week1/calculator/OperatorType.kt:
```diff
@@ -0,0 +1,36 @@
+
+enum class OperatorType(
+    val value: String,
+) : OperatorExecutor {
```
`OperatorExecutor` 인터페이스를 구현할 필요 없이 

OperatorType의 프로퍼티로 [SAM](https://kotlinlang.org/docs/fun-interfaces.html#sam-conversions)을 두고 사용하면 지금처럼 매 enum 요소에서 execute 메서드를 오버라이드할 필요기 없이 조금 더 깔끔한 코드를  만들 수 있을 것 같아요. 😉

Rok93 commented on src/main/kotlin/week1/calculator/Calculator.kt:
```diff
@@ -0,0 +1,33 @@
+    private const val DELIMITER = ' '
+
+    fun process(input: String?): Int {
+        require(!input.isNullOrBlank()) {
+            if (input == null) "식이 null입니다." else "비어 있는 식입니다."
+        }
```
예외 메시지를 분기처리하여 출력하는 것보다는 아래와 같은 예외메시지로 변경하여 input으로 어떤 값이 들어오는지를 예외메시지에 포함시키면 왜 예외가 발생했는지 원인 파악이 쉬울 것 같아요. 😃
```suggestion
        require(!input.isNullOrBlank()) {  "입력 값은 null 또는 빈 문자열을 허용하지 않습니다. input: $input" }
```

Rok93 commented on src/main/kotlin/week1/calculator/Calculator.kt:
```diff
@@ -0,0 +1,33 @@
+
+    private const val DELIMITER = ' '
+
+    fun process(input: String?): Int {
```
process라는 네이밍은 메서드의 동작을 직관적으로 표현하지 못하는 것 같아요. 🤔
<img width="817" alt="image" src="https://github.com/next-step/kotlin-racingcar/assets/47850258/9a8af005-aa63-47d1-bca5-db20648fb39c">

`calcualte`와 같이 명확한 네이밍으로 변경해볼 수 있을 것 같아요. 
또한 메서드 파라미터 또한 Calculator 객체 입장에서는 이게 입력 값인지 아닌지는 관심사가 아니기 때문에 `input`이라는 네이밍보다는 조금 더 Calculator 객체가 이 메서드 파라미터의 연관관계를 잘 표현해줄 수 있는 네이밍이 좋을 것 같습니다. 👀

메서드 파라미터로 받은 값은 `수식`이니 수식이라는 네이밍으로 변경해보면 어떨까요? 🤗

Rok93 reviewed (approved):
보성님 피드백 반영 잘해주셨네요! 😃
몇몇 피드백 남겨두었고 3단계 진행하면서 반영해주셔도 상관없을 것 같아서 
2단계는 이만 머지 하겠습니다. 🤗

3단계에서 뵐게요! 🙏
미션 진행하면서 어려운 부분이 있으면 언제든지 DM 주세요. 😉

Rok93 commented on src/main/kotlin/week1/calculator/OperatorType.kt:
```diff
@@ -0,0 +1,36 @@
+package week1.calculator
+
+enum class OperatorType(
```
이제는 OperatorType 객체가 연산을 하는 책임을 가지게된 것 같아요. 👀
네이밍을 `Operator`로 변경해보면 어떨까요? 😉
